specs/interop/predeploys.md

```diff
@@ -159,6 +163,23 @@ as well as domain binding, ie the executing transaction can only be valid on a s
 - The `_destination` chain id MUST be equal to the local chain id
 - The `CrossL2Inbox` cannot call itself
 
+### `sendMessageHashBack` Invariants
+
+- The Source chain id MUST not be `block.chainid`
+- The `_destination` chain id MUST be equal to the local chain id
+- The message MUST have not been successfully relayed
+- The `RETURN_DELAY` MUST have elapsed since the message first failed to be relayed
+- The sent-back message MUST not have been previously relayed
+- The sent-back message MUST not be relayable after being sent back
+
+### `receiveMessageHash` Invariants
+
+- Only callable by the `CrossL2Inbox`
+- The message source MUST be `block.chainid`
+- The sender MUST be `address(L2ToL2CrossDomainMessenger)`
+- The `Identifier.origin` MUST be `address(L2ToL2CrossDomainMessenger)`
+- The `Identifier.chainId` MUST be `_destination`
+
```

Comment:
If a message is sent back, we would want an invariant such that it cannot be replayed again?

Reply:
yes! this is done by setting the message as successful on the reverted domain `successfulMessages[messageHash] = true`.
we have the invariant: `- The sent-back message MUST not be relayable after being sent back` under `sendMessageHashBack Invariants`